google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITForeignKeyDeleteCascadeTest.java

```diff
@@ -0,0 +1,540 @@
+      while (rs.next()) {
+        assertThat(rs.getString(DELETE_RULE_COLUMN_NAME)).isEqualTo(DELETE_RULE_CASCADE);
+      }
+    }
+  }
+
+  @Test
+  public void testForeignKeyDeleteCascadeConstraints_withAlterDDLStatements() throws Exception {
+    // Creating new tables within this test to ensure we don't pollute tables used by other tests in this class.
+    final List<String> createStatements = getCreateAndAlterTableStatementsWithForeignKey();
+    final Database createdDatabase = createDatabase(createStatements);
+    dbs.add(createdDatabase);
+
+    final DatabaseClient databaseClient = getCreatedDatabaseClient();
+
+    final String referentialConstraintQuery = getReferentialConstraintsQueryStatement();
+    try (final ResultSet rs =
+        databaseClient
+            .singleUse()
+            .executeQuery(
+                Statement.of(referentialConstraintQuery))) {
+      while (rs.next()) {
+        assertThat(rs.getString(DELETE_RULE_COLUMN_NAME)).isEqualTo(DELETE_RULE_CASCADE);
+      }
+    }
+
+    // remove the foreign key delete cascade constraint
+    final List<String> alterDropStatements = getAlterDropForeignKeyDeleteCascadeStatements();
+    getDatabaseAdminClient()
+        .updateDatabaseDdl(env.getTestHelper().getInstanceId().getInstance(),
+            createdDatabase.getId().getDatabase(), alterDropStatements, null)
+        .get();
+
+    try (final ResultSet rs =
+        databaseClient
+            .singleUse()
+            .executeQuery(Statement.of(getReferentialConstraintsQueryStatementV2()))) {
+      while (rs.next()) {
+        assertThat(rs.getString(DELETE_RULE_COLUMN_NAME)).isEqualTo(DELETE_RULE_DEFAULT);
+      }
+    }
+  }
+
+  @Test
+  public void testForeignKeyDeleteCascadeConstraints_verifyValidInsertions() {
+
+    final DatabaseClient databaseClient = getCreatedDatabaseClient();
+    final String singerInsertStatement = getInsertStatementForSingerTable();
+    final Statement singerInsertStatementWithValues = Statement.newBuilder(singerInsertStatement)
+        // Use 'p1' to bind to the parameter with index 1 etc.
+        .bind("p1").to(1L)
+        .bind("p2").to("singerName").build();
+
+    final String concertInsertStatement = getInsertStatementForConcertsTable();
+    final Statement concertInsertStatementWithValues = Statement.newBuilder(concertInsertStatement)
+        // Use 'p1' to bind to the parameter with index 1 etc.
+        .bind("p1").to(1L)
+        .bind("p2").to(1L).build();
+
+    // successful inserts into referenced and referencing tables
+    databaseClient
+        .readWriteTransaction()
+        .run(
+            transaction -> {
+              transaction.executeUpdate(singerInsertStatementWithValues);
+              return null;
+            });
+    databaseClient
+        .readWriteTransaction()
+        .run(
+            transaction -> {
+              transaction.executeUpdate(concertInsertStatementWithValues);
+              return null;
+            });
```

Comment:
Earlier I wasn't sure if the order would be maintained in batchUpdate since due to FK relationship we'd want singer inserts to execute before concert inserts.

After reading documentation of `batchUpdate` method, I think we can have a single transaction.